Fix RDS Instance Password Reset Loop

PocketMobsters opened the pull request:
<!--
Thank you for helping to improve Crossplane!

We strongly recommend you look through our contributor guide at https://git.io/fj2m9
if this is your first time opening a Crossplane pull request. You can find us in
https://slack.crossplane.io/messages/dev if you need any help contributing.
-->

### Description of your changes
<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

Fixes #
-->
Fixes #329 

RDSInstances will now only update the master password if there's a difference between the exported secret (if exists) and the referenced input secret from `masterPasswordSecretRef`.
### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the dependencies in [`app.yaml`] to include any new role permissions.

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplane/crossplane/tree/master/PendingReleaseNotes.md
[`app.yaml`]: https://github.com/crossplane/provider-aws/blob/master/config/package/manifests/app.yaml

PocketMobsters commented:
Tacking on some additional fields to ignore in IsUpToDate. 

ApplyModificationsImmediately will cause this function to always return true because it is not returned by AWS. If this is set, it WILL be included in any subsequent modify call so it won't break any behavior as any other field that should be sent out will appropriately return true from this function.

Similarly with AllowMajorVersionUpgrade it is not returned by AWS but their [docs](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.Upgrading.html#USER_UpgradeDBInstance.Upgrading.Manual) mention that it should be included with a call that changes EngineVersion. By itself, it should not trigger a modify call to be sent. 

These two fields will ALSO cause the operator to go into a continual loop of processing these instances which is only really noticeable if you have debug logging enabled. Each time this happens, it is reporting that there were no modifications included in the request.

muvaf commented on pkg/controller/database/rdsinstance.go:
```diff
@@ -89,11 +89,29 @@ type external struct {
 		return managed.ExternalObservation{}, errors.New(errNotRDSInstance)
 	}
+
+	pwdChanged := false
```
I believe this check belongs to `IsUpToDate` function. As you said, we don't have another resource whose `IsUpToDate` func has a client but we don't have another resource in AWS that takes a secret ref as input either, so this could be example to future implementations.

I would suggest implementing this functionality in a way to be called like `pw, changed, err := rds.CheckPassword(...)` so that we can call it in `IsUpToDate`, `Update` and `Create`, essentially wherever we need the password. `pkg/clients/rds/rds.go` could be a good fit for this utility.

muvaf commented on pkg/clients/rds/rds.go:
```diff
@@ -452,6 +445,9 @@ func IsUpToDate(p v1beta1.RDSInstanceParameters, db rds.DBInstance) (bool, error
 		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "Tags"),
 		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "SkipFinalSnapshotBeforeDeletion"),
 		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "FinalDBSnapshotIdentifier"),
+		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "MasterPasswordSecretRef"),
```
Thanks for catching these!

muvaf reviewed (commented):
Hi @PocketMobsters ! Your solution looks solid, I just left a comment about where it could go. Thank you for catching this bug and opening the PR!

kelvinwijaya commented:
Thanks @PocketMobsters , i think the fix would help my existing use case for updating RDSInstances password by secret object reference (As per my current understanding the password is auto-generated), i did some workaround to run another controller for updating the RDSInstances password, but seem i don't have to anymore

PocketMobsters commented:
@muvaf I rebased my changes so everything should be up to date with master as of the time of this comment.

I moved the password change detection logic to the RDS client and thus I'm passing some additional information (context and kubernetes client) into `IsUpToDate` so it can be used in `GetPassword`. I really liked your suggestion of the return signature being `pw, changed, err`. 

After I made my changes I realized you would have named this function `CheckPassword` and I'm totally down to rename it, just want to get a bit of a review on the new approach and whether its something you'd like in the code base.

I was having some trouble coming up with a good way to test the new functionality, as I didn't really notice anything existing that mocks external calls and the contribution guidelines specifically prohibit adding a third party testing library. As it stands now, my code is untested programmatically - I've just put it through a bunch of manual integration tests. I looked at the existing E2E integration test and it appears to revolve around making sure the provider itself can be installed and brings up the appropriate deployments and pods so I don't think its appropriate for this either. If you have any ideas or existing examples of integration tests like that I'd be happy to revise my changes accordingly!

muvaf commented on pkg/clients/rds/rds.go:
```diff
@@ -435,14 +444,11 @@ func LateInitialize(in *v1beta1.RDSInstanceParameters, db *rds.DBInstance) { //
-
-	// TODO(muvaf): If a secret is provided for password, this logic should check
-	// whether it's changed by comparing it to the password in the published secret.
+func IsUpToDate(ctx context.Context, kube *client.Client, p v1beta1.RDSInstanceParameters, db rds.DBInstance, outCreds *v1alpha1.SecretReference) (bool, error) {
```
```suggestion
func IsUpToDate(ctx context.Context, kube *client.Client, p v1beta1.RDSInstance, db rds.DBInstance) (bool, error) {
```

What do you think about supplying the whole `RDSInstance` and get the necessart secret refs in the function?

muvaf replied:
```suggestion
func IsUpToDate(ctx context.Context, kube client.Client, p v1beta1.RDSInstanceParameters, db rds.DBInstance, outCreds *v1alpha1.SecretReference) (bool, error) {
```

Since `client.Client` is an interface, we don't need to send it via pointer.

muvaf commented on pkg/clients/rds/rds.go:
```diff
@@ -452,7 +458,59 @@ func IsUpToDate(p v1beta1.RDSInstanceParameters, db rds.DBInstance) (bool, error
+}
+
+// GetPassword fetches the referenced input password for an RDSInstance CRD and determines whether it has changed or not
+func GetPassword(ctx context.Context, kube *client.Client, inputCreds *v1alpha1.SecretKeySelector, outCreds *v1alpha1.SecretReference) (string, bool, error) {
```
```suggestion
func GetPassword(ctx context.Context, kube client.Client, rds *v1beta1.RDSInstance) (string, bool, error) {
```

Similar suggestions as above.

muvaf commented on pkg/clients/rds/rds.go:
```diff
@@ -452,7 +458,59 @@ func IsUpToDate(p v1beta1.RDSInstanceParameters, db rds.DBInstance) (bool, error
+	var err error
+	var newPwd string
+	var pwdChanged bool
+	if inputCreds != nil {
```
I believe once we accept `*v1beta1.RDSInstance` and do the necessary checks like `if rds.Spec.ForProvider.MasterPasswordSecretRef == nil` in the beginning, we don't need to check them anymore. Once they are out of the way, it could be a simple flow like `do the checks -> get the input secret -> get the output secret -> return newPwd, newPwd != oldPwd, nil`

PocketMobsters replied:
I basically just exit early if neither are set. It's kind of a redundant conditional upon further inspection.

I need to check each one specifically, because otherwise accessing the name, namespace and key (for input) will throw an error. If you take a look at the updated function and can figure out a less verbose way to write it and cover the cases it does I'll happily make changes!

muvaf replied:
I left some suggestions about this.

muvaf commented on pkg/clients/rds/rds.go:
```diff
@@ -452,7 +458,59 @@ func IsUpToDate(p v1beta1.RDSInstanceParameters, db rds.DBInstance) (bool, error
+	return newPwd, pwdChanged, nil
+}
+
+func getPasswordFromSecret(ctx context.Context, kube *client.Client, namespacedName types.NamespacedName, key string) (string, error) {
```
From what I can tell, once we use `client.Client`, this function would not be really necessary and we can do the same operation in place in `GetPassword`.

muvaf commented on pkg/controller/database/rdsinstance.go:
```diff
@@ -204,18 +205,16 @@ func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.Ext
 	modify := rds.GenerateModifyDBInstanceInput(meta.GetExternalName(cr), patch)
 	var conn managed.ConnectionDetails
 	if cr.Spec.ForProvider.MasterPasswordSecretRef != nil {
```
I think we can skip the `if cr.Spec.ForProvider.MasterPasswordSecretRef != nil` here since we do it in `GetPassword` anyway. The `if changed` condition covers the case where there is no change.

muvaf commented on pkg/controller/database/rdsinstance.go:
```diff
@@ -204,18 +205,16 @@ func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.Ext
-		modify.MasterUserPassword = aws.String(string(s.Data[cr.Spec.ForProvider.MasterPasswordSecretRef.Key]))
 	}
 	if _, err = e.client.ModifyDBInstanceRequest(modify).Send(ctx); err != nil {
 		return managed.ExternalUpdate{}, errors.Wrap(err, errModifyFailed)
```
Would you be use `GetPassword` in `Create`, too? We wouldn't really check whether password changed but we'd need the desired password and generate a new one only if the returned password is `""`.

PocketMobsters replied:
Wonderful comments, I'll make the according changes and look at the tests you referenced to see what I can do with the clients!

muvaf reviewed (commented):
Thank you for the update! It's looking pretty good! I've left a few simplification suggestions.

Regarding tests, I see that `pkg/client/rds/rds_test.go` file assumes that all functions are bare conversions, hence, doesn't have any test with clients. But we do have such tests and you can find some examples in [`pkg/controller/database/rdsinstance_test.go`](https://github.com/crossplane/provider-aws/blob/master/pkg/controller/database/rdsinstance_test.go#L232) and in other controller tests.

PocketMobsters commented:
I've successfully forgotten to sign off my commits at least 3 times here. One day I'll get it done without DCO reminding me.

@muvaf I believe I've addressed most of your comments, with the exception being the flow change in `GetPassword` because I'm not 100% sure how to change it to work properly without the individual checks. I think some of the confusion comes from the early exit if neither are set. 

I'd like to apologize for the delay on getting these changes in, I've been pulling on a couple different threads at work in the meantime.

muvaf commented on pkg/clients/rds/rds.go:
```diff
@@ -452,7 +458,50 @@ func IsUpToDate(p v1beta1.RDSInstanceParameters, db rds.DBInstance) (bool, error
+
+// GetPassword fetches the referenced input password for an RDSInstance CRD and determines whether it has changed or not
+func GetPassword(ctx context.Context, kube client.Client, r *v1beta1.RDSInstance) (string, bool, error) {
+	if kube == nil {
+		return "", false, nil
+	}
```
```suggestion
```

I don't think we need to check this because if that's `nil`, probably many of the calls before this function will fail anyway.

muvaf commented on pkg/clients/rds/rds.go:
```diff
@@ -452,7 +458,50 @@ func IsUpToDate(p v1beta1.RDSInstanceParameters, db rds.DBInstance) (bool, error
+	if kube == nil {
+		return "", false, nil
+	}
+	if r.Spec.ForProvider.MasterPasswordSecretRef == nil && r.Spec.WriteConnectionSecretToReference == nil {
```
```suggestion
	if r.Spec.ForProvider.MasterPasswordSecretRef == nil || r.Spec.WriteConnectionSecretToReference == nil {
```

In order for this function to make its job, both fields have to be set. So, if any of them isn't there, then we should return early. As written here, that's not an error case, it's just saying that we don't need the utility of `GetPassword` for those cases.

muvaf replied:
After you make this change, in the rest of the function you can be sure that they are not `nil` so you can get rid of two `if` conditions below.

PocketMobsters replied:
Neither `masterPasswordSecretRef` (input) nor `writeConnectionSecretToRef` (output) are marked as required, so I figure we should try and account for situations where either are set. I figure the best way to dissect this would be to walk through the contexts in which this function may be called.

-----

### Create

When creating an instance, we only really check the input. If input is set, we'll use that value for the password we provide to AWS. If the input field is not set, or the value in the referenced secret is set to `""` we'll autogenerate a password. I've created a small diagram for [it here](https://i.imgur.com/Qn7BJTL.png).

If we change the `GetPassword` function to return if either input or output references are nil, we wouldn't report back what the input password is so I'd revert the change in `Create` and have it pull the input secret directly.

### Update

I created [another diagram](https://i.imgur.com/kEfYaRR.png) to help convey my thought process. As you can see, we don't really do anything unless both fields are set, so I can see how making this change would clean up the code a bit.

### Observe

I created [a diagram](https://i.imgur.com/pPM4rCz.png) to visually represent this flow as well. Again we only really do work if both fields are set.

-----

If we have `Create` pull the password directly as it was doing before, I can 100% support cleaning up the code a bit. I'd also feel like a rename from `GetPassword` to `CheckPassword` as suggested in your original round of reviews would completely describe what it is doing. Otherwise, I feel we should at least continue supporting fields being partially set, to account for the `Create` case.

muvaf replied:
Thanks for all the details! I think I understand what you're describing. Seems like we have only one case that's tricky; input is given but output is not given during `Create`. I couldn't really think of many scenarios where you give an input secret (meaning you use pwd auth) but not use the output. However, for the implementation to be _correct_, i.e. user supplied a password and that should be enough for us to use it, I agree with you that this should be taken into account.

So, in all cases, if input is empty we should return early but output being empty is not of a concern so that we can take `Create` into account. So, I'd suggest:
```suggestion
	if r.Spec.ForProvider.MasterPasswordSecretRef == nil {
```

and assume that `r.Spec.ForProvider.MasterPasswordSecretRef` is not `nil` in the rest of the function. The check for `if r.Spec.WriteConnectionSecretToReference != nil {` in the later parts of the function seems necessary.

PocketMobsters replied:
Pushed this up

muvaf commented on pkg/clients/rds/rds.go:
```diff
@@ -452,7 +458,50 @@ func IsUpToDate(p v1beta1.RDSInstanceParameters, db rds.DBInstance) (bool, error
+			Namespace: r.Spec.WriteConnectionSecretToReference.Namespace,
+		}
+		s := &corev1.Secret{}
+		if err := kube.Get(ctx, nn, s); err != nil {
+			pwdChanged = newPwd != ""
+		} else {
+			outPwd := string(s.Data[v1alpha1.ResourceCredentialsSecretPasswordKey])
+			pwdChanged = newPwd != "" && outPwd != newPwd
+		}
```
```suggestion
		if err := kube.Get(ctx, nn, s); resource.IgnoreNotFound(err) != nil {
			return "", false, err
		}
		pwdChanged = newPwd != string(s.Data[v1alpha1.ResourceCredentialsSecretPasswordKey])
		err = nil
		return
```

On `NotFound` case is OK to skip. For other errors, we need to return them. Also we don't need to check specifically against `""` because if it's `NotFound`, then `string(s.Data[v1alpha1.ResourceCredentialsSecretPasswordKey])` will be `""`.

PocketMobsters replied:
Great call

PocketMobsters replied:
I kept the `newPwd != ""` because I think we want to make sure that at least some value was pulled from the input secret, otherwise we may report that the password changed (a value in output but no value in input) but also send back an empty string for `newPwd`. This wouldn't matter in `Create` as we only care about the value of the password, but I'm unsure what the implications are of providing an empty password in `Update` and what having that included in the AWS request body would do

muvaf commented on pkg/clients/rds/rds.go:
```diff
@@ -452,7 +458,50 @@ func IsUpToDate(p v1beta1.RDSInstanceParameters, db rds.DBInstance) (bool, error
+}
+
+// GetPassword fetches the referenced input password for an RDSInstance CRD and determines whether it has changed or not
+func GetPassword(ctx context.Context, kube client.Client, r *v1beta1.RDSInstance) (string, bool, error) {
```
```suggestion
func GetPassword(ctx context.Context, kube client.Client, r *v1beta1.RDSInstance) (newPwd string, pwdChanged bool, err error) {
```

I think naming the returned variables if there are more than two would serve as a nice documentation to understand what the `bool` is for.

muvaf commented:
> I'd like to apologize for the delay on getting these changes in, I've been pulling on a couple different threads at work in the meantime.

No need to apologize, it's perfectly fine. We wouldn't even know about this problem. Thank you for driving this, it's pretty close to be merged!

muvaf reviewed (approved):
LGTM! Thank you @PocketMobsters for working on this!
